### ISSUE TITLE
Timer auto-reload register miscalculation

### ISSUE THREAD

**Issue by albru123** [stm32-rs/stm32f4xx-hal#260]
After poking into timer configuration and staring into the RM for a while I think I found a slight mistake in the timer configuration code. The auto-reload register divides the timer frequency the same way prescaler does, so the division factor is actually `(ARR + 1)`.

Here's the line:
https://github.com/stm32-rs/stm32f4xx-hal/blob/9b6aad4b3365a48ae652c315730ab47522e57cfb/src/timer.rs#L314

I think it should look like this:
```rust
let arr = u16((ticks / u32(psc + 1)) - 1).unwrap();
```

Could someone please look into this and tell me whether I'm right or not?

Maybe @aurelj might have a clue.

**Comment by David-OConnor** [comment]
You're correct; this is a bug that's been copy+pasted into a number of HALs.
